app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,174 @@
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * Allows evaluation contexts to support the here() XPath function,
+ * which returns the current location.
+ *
+ * In addition, an EntitySelectActivity can register itself to be refreshed whenever
+ * a new value of here() is obtained (whenever the location changes).
+ *
+ * No locations are requested if here() is never evaluated.
+ *
+ * @author Forest Tong
+ */
+public class HereFunctionHandler implements IFunctionHandler, LocationListener {
+    public static final String HERE_NAME = "here";
+    private GeoPointData location;
+
+    private LocationManager mLocationManager;
+
+    private boolean requestingLocationUpdates;
+    private boolean locationGoodEnough;
+
+    private final Context context = CommCareApplication._().getApplicationContext();
+
+    // If there are more general uses for HereFunctionHandler, the type of this field can be
+    // generalized to a listener interface.
+    private EntitySelectActivity entitySelectActivity;
+
+    public HereFunctionHandler() {}
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    public void unregisterEntitySelectActivity() {
+        this.entitySelectActivity = null;
+    }
+
+    // The EntitySelectActivity must subscribe before this method is called if a fresh location is desired.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if (entitySelectActivity != null && !entitySelectActivity.containsHereFunction) {
+            refreshLocation();
+            allowGpsUse();
+            entitySelectActivity.containsHereFunction = true;
+        }
+        if (location == null) {
+            return "";
+        }
+        return location.getDisplayText();
+    }
+
+    public void allowGpsUse() {
+        if (!locationGoodEnough && !requestingLocationUpdates) {
+            requestLocationUpdates();
+        }
+    }
+
+    public void forbidGpsUse() {
+        if (requestingLocationUpdates) {
+            removeLocationUpdates();
+        }
+    }
+
+    public void refreshLocation() {
+        this.locationGoodEnough = false;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        this.location = toGeoPointData(location);
+        Log.i("HereFunctionHandler", "location has been set to " + this.location.getDisplayText());
+
+        if (this.location.getAccuracy() <= GeoUtils.ACCEPTABLE_ACCURACY) {
```

Comment:
`getAccuracy` isn't in the current master branch of javarosa. Be sure to create a cross-requested javarosa PR so when this gets merged it doesn't break everyone's master builds.